File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> v;
        if (root == NULL) return v;

        queue<TreeNode*> q;
        q.push(root);

        int flag = 0;

        while (!q.empty()) {
            int size = q.size();
            vector<int> level(size);

            for (int i = 0; i < size; i++) {
                TreeNode* node = q.front();
                q.pop();

                // choose index based on direction
                int ind = (flag == 0) ? i : (size - 1 - i);

                level[ind] = node->val;

                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }

            v.push_back(level);
            flag = !flag; // toggle direction
        }

        return v;
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) return ans;
        q.add(root);
        boolean toggle = true;
        while (!q.isEmpty()) {
            List<Integer> level = new ArrayList<>();
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode no = q.poll();
                level.add(no.val);
                if (no.left != null)
                    q.add(no.left);
                if (no.right != null)
                    q.add(no.right);
            }

            if (!toggle)
                Collections.reverse(level);
            toggle = !toggle;
            ans.add(level);
        }
        return ans;
    }
}